### ISSUE TITLE
facts gathering fails on a dual-RE MXVC doesn't supply version of all REs

### ISSUE THREAD

**Issue by stacywsmith** [Juniper/py-junos-eznc#416]
facts dictionary only has these keys for an MXVC with 4 routing engines (2 REs per member):

 'version': '14.2R4.2',
 'version_RE0': '14.2R4.2',
 'version_RE1': '14.2R4.2',
 'version_info': junos.version_info(major=(14, 2), type=R, minor=4, build=2)}

user@h0$ python
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.

> > > from jnpr.junos import Device
> > > from pprint import pprint
> > > dev = Device(host="mxvc",
> > > ...              user='user',
> > > ...              password='user123',
> > > ...              port=22)
> > > dev.open()
> > > Device(mxvc)
> > > pprint(dev.facts)
> > > {'2RE': True,
> > >  'HOME': '/var/home/user',
> > >  'RE0-RE0': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
> > >              'mastership_state': 'master',
> > >              'model': 'RE-S-1800x4',
> > >              'status': 'OK',
> > >              'up_time': '19 hours, 22 minutes, 58 seconds'},
> > >  'RE0-RE1': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
> > >              'mastership_state': 'backup',
> > >              'model': 'RE-S-1800x4',
> > >              'status': 'OK',
> > >              'up_time': '19 hours, 22 minutes, 48 seconds'},
> > >  'RE1-RE0': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
> > >              'mastership_state': 'master',
> > >              'model': 'RE-S-1800x4',
> > >              'status': 'OK',
> > >              'up_time': '19 hours, 22 minutes, 53 seconds'},
> > >  'RE1-RE1': {'last_reboot_reason': 'Router rebooted after a normal shutdown.',
> > >              'mastership_state': 'backup',
> > >              'model': 'RE-S-1800x4',
> > >              'status': 'OK',
> > >              'up_time': '19 hours, 22 minutes, 47 seconds'},
> > >  'RE_hw_mi': True,
> > >  'domain': 'example.com',
> > >  'fqdn': 'mxvc.example.com',
> > >  'hostname': 'mxvc',
> > >  'ifd_style': 'CLASSIC',
> > >  'master': 'RE0',
> > >  'model': 'MX240',
> > >  'personality': 'MX',
> > >  'serialnumber': 'JN1230B3FAFC',
> > >  'switch_style': 'BRIDGE_DOMAIN',
> > >  'vc_capable': True,
> > >  'vc_mode': None,
> > >  'version': '14.2R4.2',
> > >  'version_RE0': '14.2R4.2',
> > >  'version_RE1': '14.2R4.2',
> > >  'version_info': junos.version_info(major=(14, 2), type=R, minor=4, build=2)}
> > > quit()
> > > user@h0$ 

PyEZ is executing 'show version all-members' on a MXVC with dual-REs, but this command only shows the master RE on each member.

{master:member0-re0}
user@mxvc> show version all-members                      
## member0:

Hostname: mxvc
Model: mx240
Junos: 14.2R4.2
JUNOS Base OS boot [14.2R4.2]
JUNOS Base OS Software Suite [14.2R4.2]
JUNOS platform Software Suite [14.2R4.2]
JUNOS Web Management [14.2R4.2]
JUNOS Runtime Software Suite [14.2R4.2]
JUNOS Online Documentation [14.2R4.2]
JUNOS Services Application Level Gateways [14.2R4.2]
JUNOS Services AACL Container package [14.2R4.2]
JUNOS Border Gateway Function package [14.2R4.2]
JUNOS AppId Services [14.2R4.2]
JUNOS IDP Services [14.2R4.2]
JUNOS Services HTTP Content Management package [14.2R4.2]
JUNOS Services Captive Portal and Content Delivery Container package [14.2R4.2]
JUNOS Services LL-PDF Container package [14.2R4.2]
JUNOS Services Jflow Container package [14.2R4.2]
JUNOS Services NAT [14.2R4.2]
JUNOS Services Mobile Subscriber Service Container package [14.2R4.2]
JUNOS Services MobileNext Software package [14.2R4.2]
JUNOS Services RPM [14.2R4.2]
JUNOS Services PTSP Container package [14.2R4.2]
JUNOS Services Stateful Firewall [14.2R4.2]
JUNOS Voice Services Container package [14.2R4.2]
JUNOS Macsec Software Suite [14.2R4.2]
JUNOS Services Crypto [14.2R4.2]
JUNOS Services SSL [14.2R4.2]
JUNOS Services IPSec [14.2R4.2]
JUNOS py-base-i386 [14.2R4.2]
JUNOS 64-bit Kernel Software Suite [14.2R4.2]
JUNOS Crypto Software Suite [14.2R4.2]
JUNOS 64-bit Runtime Software Suite [14.2R4.2]
JUNOS Packet Forwarding Engine Support (M/T/EX Common) [14.2R4.2]
JUNOS Packet Forwarding Engine Support (MX Common) [14.2R4.2]
JUNOS Routing Software Suite [14.2R4.2]
## member1:                                

Hostname: mxvcb
Model: mx240
Junos: 14.2R4.2
JUNOS Base OS boot [14.2R4.2]
JUNOS Base OS Software Suite [14.2R4.2]
JUNOS platform Software Suite [14.2R4.2]
JUNOS Web Management [14.2R4.2]
JUNOS Runtime Software Suite [14.2R4.2]
JUNOS Online Documentation [14.2R4.2]
JUNOS Services AACL Container package [14.2R4.2]
JUNOS AppId Services [14.2R4.2]
JUNOS Services Application Level Gateways [14.2R4.2]
JUNOS Border Gateway Function package [14.2R4.2]
JUNOS Services HTTP Content Management package [14.2R4.2]
JUNOS Services Captive Portal and Content Delivery Container package [14.2R4.2]
JUNOS IDP Services [14.2R4.2]
JUNOS Services Jflow Container package [14.2R4.2]
JUNOS Services LL-PDF Container package [14.2R4.2]
JUNOS Services MobileNext Software package [14.2R4.2]
JUNOS Services NAT [14.2R4.2]
JUNOS Services Mobile Subscriber Service Container package [14.2R4.2]
JUNOS Services PTSP Container package [14.2R4.2]
JUNOS Services RPM [14.2R4.2]
JUNOS Services Stateful Firewall [14.2R4.2]
JUNOS Voice Services Container package [14.2R4.2]
JUNOS Services Crypto [14.2R4.2]
JUNOS Macsec Software Suite [14.2R4.2]
JUNOS Services SSL [14.2R4.2]
JUNOS Services IPSec [14.2R4.2]
JUNOS py-base-i386 [14.2R4.2]
JUNOS 64-bit Kernel Software Suite [14.2R4.2]
JUNOS Crypto Software Suite [14.2R4.2]
JUNOS 64-bit Runtime Software Suite [14.2R4.2]
JUNOS Packet Forwarding Engine Support (M/T/EX Common) [14.2R4.2]
JUNOS Packet Forwarding Engine Support (MX Common) [14.2R4.2]
JUNOS Routing Software Suite [14.2R4.2]

{master:member0-re0}
user@mxvc> 

The info from all 4 REs can be gathered with "show version all-members invoke-on all-routing-engines" or just "show version invoke-on all-routing-engines"

{master:member0-re0}
user@mxvc> show version all-members invoke-on all-routing-engines           
## member0-re0:

Hostname: mxvc
Model: mx240
Junos: 14.2R4.2
JUNOS Base OS boot [14.2R4.2]
JUNOS Base OS Software Suite [14.2R4.2]
JUNOS platform Software Suite [14.2R4.2]
JUNOS Web Management [14.2R4.2]
JUNOS Runtime Software Suite [14.2R4.2]
JUNOS Online Documentation [14.2R4.2]
JUNOS Services Application Level Gateways [14.2R4.2]
JUNOS Services AACL Container package [14.2R4.2]
JUNOS Border Gateway Function package [14.2R4.2]
JUNOS AppId Services [14.2R4.2]
JUNOS IDP Services [14.2R4.2]
JUNOS Services HTTP Content Management package [14.2R4.2]
JUNOS Services Captive Portal and Content Delivery Container package [14.2R4.2]
JUNOS Services LL-PDF Container package [14.2R4.2]
JUNOS Services Jflow Container package [14.2R4.2]
JUNOS Services NAT [14.2R4.2]
JUNOS Services Mobile Subscriber Service Container package [14.2R4.2]
JUNOS Services MobileNext Software package [14.2R4.2]
JUNOS Services RPM [14.2R4.2]
JUNOS Services PTSP Container package [14.2R4.2]
JUNOS Services Stateful Firewall [14.2R4.2]
JUNOS Voice Services Container package [14.2R4.2]
JUNOS Macsec Software Suite [14.2R4.2]
JUNOS Services Crypto [14.2R4.2]
JUNOS Services SSL [14.2R4.2]
JUNOS Services IPSec [14.2R4.2]
JUNOS py-base-i386 [14.2R4.2]
JUNOS 64-bit Kernel Software Suite [14.2R4.2]
JUNOS Crypto Software Suite [14.2R4.2]
JUNOS 64-bit Runtime Software Suite [14.2R4.2]
JUNOS Packet Forwarding Engine Support (M/T/EX Common) [14.2R4.2]
JUNOS Packet Forwarding Engine Support (MX Common) [14.2R4.2]
JUNOS Routing Software Suite [14.2R4.2]
## member0-re1:                            

Hostname: mxvc1
Model: mx240
Junos: 14.2R4.2
JUNOS Base OS boot [14.2R4.2]
JUNOS Base OS Software Suite [14.2R4.2]
JUNOS platform Software Suite [14.2R4.2]
JUNOS Web Management [14.2R4.2]
JUNOS Runtime Software Suite [14.2R4.2]
JUNOS Online Documentation [14.2R4.2]
JUNOS Services Application Level Gateways [14.2R4.2]
JUNOS Services AACL Container package [14.2R4.2]
JUNOS Border Gateway Function package [14.2R4.2]
JUNOS AppId Services [14.2R4.2]
JUNOS IDP Services [14.2R4.2]
JUNOS Services HTTP Content Management package [14.2R4.2]
JUNOS Services Captive Portal and Content Delivery Container package [14.2R4.2]
JUNOS Services LL-PDF Container package [14.2R4.2]
JUNOS Services Jflow Container package [14.2R4.2]
JUNOS Services Mobile Subscriber Service Container package [14.2R4.2]
JUNOS Services MobileNext Software package [14.2R4.2]
JUNOS Services NAT [14.2R4.2]
JUNOS Services RPM [14.2R4.2]
JUNOS Services PTSP Container package [14.2R4.2]
JUNOS Services Stateful Firewall [14.2R4.2]
JUNOS Voice Services Container package [14.2R4.2]
JUNOS Macsec Software Suite [14.2R4.2]
JUNOS Services Crypto [14.2R4.2]
JUNOS Services SSL [14.2R4.2]
JUNOS Services IPSec [14.2R4.2]
JUNOS py-base-i386 [14.2R4.2]
JUNOS 64-bit Kernel Software Suite [14.2R4.2]
JUNOS Crypto Software Suite [14.2R4.2]
JUNOS 64-bit Runtime Software Suite [14.2R4.2]
JUNOS Packet Forwarding Engine Support (M/T/EX Common) [14.2R4.2]
JUNOS Packet Forwarding Engine Support (MX Common) [14.2R4.2]
JUNOS Routing Software Suite [14.2R4.2]
## member1-re0:                            

Hostname: mxvcb
Model: mx240
Junos: 14.2R4.2
JUNOS Base OS boot [14.2R4.2]
JUNOS Base OS Software Suite [14.2R4.2]
JUNOS platform Software Suite [14.2R4.2]
JUNOS Web Management [14.2R4.2]
JUNOS Runtime Software Suite [14.2R4.2]
JUNOS Online Documentation [14.2R4.2]
JUNOS Services AACL Container package [14.2R4.2]
JUNOS AppId Services [14.2R4.2]
JUNOS Services Application Level Gateways [14.2R4.2]
JUNOS Border Gateway Function package [14.2R4.2]
JUNOS Services HTTP Content Management package [14.2R4.2]
JUNOS Services Captive Portal and Content Delivery Container package [14.2R4.2]
JUNOS IDP Services [14.2R4.2]
JUNOS Services Jflow Container package [14.2R4.2]
JUNOS Services LL-PDF Container package [14.2R4.2]
JUNOS Services MobileNext Software package [14.2R4.2]
JUNOS Services NAT [14.2R4.2]
JUNOS Services Mobile Subscriber Service Container package [14.2R4.2]
JUNOS Services PTSP Container package [14.2R4.2]
JUNOS Services RPM [14.2R4.2]
JUNOS Services Stateful Firewall [14.2R4.2]
JUNOS Voice Services Container package [14.2R4.2]
JUNOS Services Crypto [14.2R4.2]
JUNOS Macsec Software Suite [14.2R4.2]
JUNOS Services SSL [14.2R4.2]
JUNOS Services IPSec [14.2R4.2]
JUNOS py-base-i386 [14.2R4.2]
JUNOS 64-bit Kernel Software Suite [14.2R4.2]
JUNOS Crypto Software Suite [14.2R4.2]
JUNOS 64-bit Runtime Software Suite [14.2R4.2]
JUNOS Packet Forwarding Engine Support (M/T/EX Common) [14.2R4.2]
JUNOS Packet Forwarding Engine Support (MX Common) [14.2R4.2]
JUNOS Routing Software Suite [14.2R4.2]
## member1-re1:                            

Hostname: mxvcb1
Model: mx240
Junos: 14.2R4.2
JUNOS Base OS boot [14.2R4.2]
JUNOS Base OS Software Suite [14.2R4.2]
JUNOS platform Software Suite [14.2R4.2]
JUNOS Web Management [14.2R4.2]
JUNOS Runtime Software Suite [14.2R4.2]
JUNOS Online Documentation [14.2R4.2]
JUNOS Services Application Level Gateways [14.2R4.2]
JUNOS Services AACL Container package [14.2R4.2]
JUNOS AppId Services [14.2R4.2]
JUNOS Border Gateway Function package [14.2R4.2]
JUNOS Services HTTP Content Management package [14.2R4.2]
JUNOS Services Captive Portal and Content Delivery Container package [14.2R4.2]
JUNOS IDP Services [14.2R4.2]
JUNOS Services Jflow Container package [14.2R4.2]
JUNOS Services LL-PDF Container package [14.2R4.2]
JUNOS Services Mobile Subscriber Service Container package [14.2R4.2]
JUNOS Services MobileNext Software package [14.2R4.2]
JUNOS Services NAT [14.2R4.2]
JUNOS Services PTSP Container package [14.2R4.2]
JUNOS Services RPM [14.2R4.2]
JUNOS Services Stateful Firewall [14.2R4.2]
JUNOS Voice Services Container package [14.2R4.2]
JUNOS Services Crypto [14.2R4.2]
JUNOS Macsec Software Suite [14.2R4.2]
JUNOS Services SSL [14.2R4.2]
JUNOS Services IPSec [14.2R4.2]
JUNOS py-base-i386 [14.2R4.2]
JUNOS 64-bit Kernel Software Suite [14.2R4.2]
JUNOS Crypto Software Suite [14.2R4.2]
JUNOS 64-bit Runtime Software Suite [14.2R4.2]
JUNOS Packet Forwarding Engine Support (M/T/EX Common) [14.2R4.2]
JUNOS Packet Forwarding Engine Support (MX Common) [14.2R4.2]
JUNOS Routing Software Suite [14.2R4.2]

{master:member0-re0}
user@mxvc>

**Comment by jamesboswell** [comment]
Modifying [line 85 of swver.py](https://github.com/Juniper/py-junos-eznc/blob/861ae0a9c4c61e928407146710ad248b1791e04f/lib/jnpr/junos/facts/swver.py#L85) 
 
*from:*
`return dev.rpc.cli("show version all-members", format='xml')`
*to:*
`return dev.rpc.cli("show version all-members invoke-on all-routing-engines", format='xml')`

should fix this.    @stacywsmith   do you have a VC system you can test this on?

**Comment by jamesboswell** [comment]
@stacywsmith `git clone -b issue416 https://github.com/jamesboswell/py-junos-eznc.git` if you can test

**Comment by spidercensus** [comment]
Wow I'm surprised we're using the cli method there.  Jeremy used to have a pretty strong bias against using that function if it was at all possible to use the XML function or one of his decorators.

Do you have a pull request in yet? I may want to change that to bring it in line with the rest of the code base.

- - - - - - - - - -
Jason Pack

On Dec 14, 2015, at 20:45, "jamesboswell" <notifications@github.com<mailto:notifications@github.com>> wrote:


@stacywsmith<https://github.com/stacywsmith> git clone -b issue416 https://github.com/jamesboswell/py-junos-eznc.git if you can test

-

**Comment by stacywsmith** [comment]
@jamesboswell
> Modifying line 85 of swver.py
>  should fix this.
> from:
```
return dev.rpc.cli("show version all-members", format='xml')
```
> to:
```
return dev.rpc.cli("show version all-members invoke-on all-routing-engines", format='xml')
```

Unfortunately, it's not that simple. There are some VC-capable systems which support the "all-members" argument, but do not support the "invoke-on all-routing-engines" argument. I believe an EX4200 VC is one such example. You're suggestion would break these systems.

There's also some buginess in the setting of facts['vc_capable'] where I don't believe the fact is always set correctly.

I also haven't had a chance to look at it in detail, but I think there might some cases where lines 140 to 172 of swver.py don't always do the correct parsing of the info which is returned.

@spidercensus 
> Wow I'm surprised we're using the cli method there.  Jeremy used to have a pretty strong bias
> against using that function if it was at all possible to use the XML function or one of his decorators.

I believe there is a reason that
```
dev.rpc.cli("show version all-members", format='xml')
```
was used instead of
```
dev.rpc.get_software_information(all_members=True)
```
I believe there were (still are?) some platforms and versions of Junos which didn't correctly support the <all-members/> argument to the <get-software-information> XML RPC, but did correctly parse the CLI command "show version all-members".

**Comment by stacywsmith** [comment]
This is addressed by #638

**Comment by stacywsmith** [comment]
#638 provides this enhancement.
